registry/registry.go

```diff
@@ -218,17 +219,24 @@ func (c *client) tagsToRepository(cancel func(), client backend, lookupName, ima
 		err   error
 	}
 
+	toFetch := make(chan string, len(tags))
 	fetched := make(chan result, len(tags))
 
-	for _, tag := range tags {
-		go func(t string) {
-			img, err := c.lookupImage(client, lookupName, imageName, t)
-			if err != nil {
-				c.Logger.Log("registry-metadata-err", err)
+	for i := 0; i < maxConcurrency; i++ {
+		go func() {
+			for tag := range toFetch {
+				img, err := c.lookupImage(client, lookupName, imageName, tag)
+				if err != nil {
+					c.Logger.Log("registry-metadata-err", err)
+				}
+				fetched <- result{img, err}
 			}
-			fetched <- result{img, err}
-		}(tag)
+		}()
+	}
+	for _, tag := range tags {
+		toFetch <- tag
 	}
+	close(toFetch)
```

Comment:
Closing a chan doesn't stop you receiving already sent values, it just stops you sending *more* values. Your examples work because you are waiting for the goroutines to finish consuming before the program exits.

But at the point we close the chan it looks like `["tag1", "tag2", "tag3", <close>]`. The worker goroutines process through all the jobs on the chan, then receive the `<close>` sigil which tells the range loop to finish.

We could have also divided up the `tags` array into `n` slices, and created a goroutine for each of them, but I thought I was more likely to write an off-by-one bug and panic that way.